<a href="https://colab.research.google.com/github/afifahnrswr/NoSQL-using-PyMongo-/blob/master/practice1_Afifah_Nur_Iswari.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practice Case Day 1

Connect cluster pymongo pada link berikut (pilih salah satu yang bisa) :
<br>
•mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority <br>
•mongodb+srv://mongodb-intro:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority
<br>
<br>
Langkah-langkah pengerjaan : <br>
**LANGKAH 1** <br>
Membuat collection baru dengan nama `clean_movies` yang sama persis dengan `movies` *collection* pada *database* `sample_mflix` dengan memakai *collection* `movies_intial` (projecting_queries pymongo) <br>
**LANGKAH 2.** <br>
Validasi collection yang telah dibuat dengan parameter sbb :
- Semua document pada `clean_movies` dan `movie` sama
- Banyak document pada `clean_movies` dan `movie` sama
- Semua fields pada `clean_movies` ada pada `movie`
- Semua value pada `clean_movies` sama dengan semua value pada `movies` dengan urutan yang sama


> Sebelum melakukan LANGKAH 1 maka dapat melakukan beberapa tahap-tahap berikut :
- install package
- import package
- import client yang akan digunakan
- melakukan definisi pada database dan collection beerdasarkan client yang telah ditentukan    

In [3]:
# install package
!pip install pymongo[srv]

     |████████████████████████████████| 194kB 10.0MB/s 


In [0]:
# import package
import pymongo
from pymongo import MongoClient

import pprint
from IPython.display import clear_output


In [0]:
# initiative connection (select one cluster)
client0 = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

In [4]:
# show all database (db) in cluster
print(client0.list_database_names())

['sample_mflix', 'admin', 'local']


In [0]:
# definisi pada database dan collection beerdasarkan client yang telah ditentukan

# define database to use
db0 = client0['sample_mflix']

# define colection to use
col01 = db0.movies
col02 = db0.movies_initial

## LANGKAH 1 : Membuat collection baru
> - *Database* yang digunakan adalah `sample_mflix` yang berada pada `client0`
- *Collection* baru yang akan dibentuk yaitu menggunakan collection `col02 = movies_initial` yang akan disesuikan berdasarkan collection `col01 = 'movies'` dan akan didefinisikan sebagai collection `clean_movies_afifah`

### 1.1 Show Collection
Sebelum membuat collection baru maka perlu diketahui terlebih dahulu `keys` atau `fields` dari kedua collection (`col01` dan `col02`)

In [20]:
# col01 = movies
pprint.pprint(col01.find_one())

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

In [21]:
# col02 = movies_initial
pprint.pprint(col02.find_one())

{'_id': ObjectId('5daf3e85ab0f2781ea55d595'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
             'dress with a hoop skirt and white high-heeled pumps, Carmencita '
             'does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
         'with a hoop skirt and white high-heeled pumps, Carmencita does a '
         'dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer':

> Untuk memudahkan dalam melakukan matching keys, maka dapat membuat dataframe dari kedua collection yang akan menjadi sumber untuk membuat collection yang baru (walaupun dataframe bukanlah fiture dari MongoDB maka untuk dapat membuat dataframe perlu melakukan import package `import pandas as pd`).

In [0]:
import pandas as pd

In [25]:
pd.DataFrame(col02.find().limit(1)) #movies initial

_id awards        cast  ...   type writer  year
0  5daf3e85ab0f2781ea55d595         Carmencita  ...  movie         1894

[1 rows x 22 columns]

In [26]:
pd.DataFrame(col01.find().limit(1)) #movies

_id  ...  year
0  573a1390f29313caabcd5a93  ...  1916

[1 rows x 19 columns]

> Berdasarkan query `find_one()` dan `find()` di atas, maka dapat diketahui terdapat 19 keys pada collection `movies`. Maka dugaan awal, collection baru yang akan dibentuk yaitu dengan menggunakan 19 keys. 

### 1.2 Know More About Movies Collection
> Untuk memastikan apakah jumlah keys pada collection `movies` telah sesuai, maka dapat dilakukan pengecekan kembali dengan memunculkan keseluruhan keys untuk beberapa baris (10 baris) pada setiap collection.

In [8]:
# keys untuk 10 baris pada collection movies
x01 = []
for doc in col01.find({}):
    x01.append(doc.keys())
x01[:10]    

[dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes']),
 dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'languages', 'released', 'directors', 'rated', 'awards', 'lastupdated', 'year', 'imdb', 'countries', 'type', 'tomatoes']),
 dict_keys(['_id', 'plot', 'genres', 'runtime', 'rated', 'cast', 'poster', 'title', 'fullplot', 'languages', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'countries', 'type', 'tomatoes']),
 dict_keys(['_id', 'plot', 'genres', 'runtime', 'rated', 'cast', 'num_mflix_comments', 'title', 'fullplot', 'languages', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'countries', 'type', 'tomatoes']),
 dict_keys(['_id', 'plot', 'genres', 'runtime', 'rated', 'cast', 'num_mflix_comments', 'poste

> Berdasarkan output di atas, maka dapat terlihat bahwa jumlah keys pada setiap baris tidak sama. Hanya saja pada saat menggunakan query `find` kebetulan hanya memunculkan 19 keys saja, hal tersebut dikarenakan query tersebut mengambil document secara random. Maka dari itu untuk setiap keys dalam pipeline akan menggunakan fungsi conditional (`$cond`) dengan syarat pada saat baris tidak mengandung keys tertentu maka keys tidak akan muncul pada document tersebut (`$nonExistinField`).

In [159]:
# keys untuk 10 baris pada collection movies_initial
x02 = []
for doc in col02.find({}):
    x02.append(doc.keys())
x02[:10]    

[dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys([

> Jika dilihat pada output di atas, maka dapat diketahui untuk jumlah keys untuk setiap baris pada collection `movies_initial` adalah sama.
<br>
<br>
Dikarenakan dalam practice kali ini ingin membuat collection baru menyerupai collection `movies` dengan keys yang berbeda beda untuk setiap barisnya (jika tidak ada value pada suatu keys, maka keys tersebut tidak akan ditampilkan pada document tersebut), berdasarkan collection `movies_initial` yang telah tersedia. Maka dari itu perlu diketahui keseluruhan keys (terpanjang)  pada collection `movies`.

In [97]:
# how to show all keys movies in one vertical list
key_movies = [list(doc.keys()) 
              for doc in col01.find({})]
key_movies = set([j 
                  for i in key_movies 
                  for j in i])
key_movies

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

> Daat diketahui bahwa terdapat 22 keys pada collection `movies` dengan nama untuk setiap keys dapat dilihat pada output di atas. Maka dari itu, dalam membuat pipeline, 22 keys tersebut perlu diimputkan pada `$project`.

In [12]:
imdb = list (col01.find({},{'_id':0, 'imdb':1}))
imdb_list = []
print ('Jumlah baris movies :', len(imdb))

Jumlah baris movies : 23541


In [10]:
print ('Jumlah baris movies_initial :', len(list(col02.find({}))))

Jumlah baris movies_initial : 46014


> Dikarenakan jumlah baris pada kedua collection berbeda, maka dari itu perlu dilakukan agregasi dengan memastikan jumlah baris pada collection yang akan dibentuk sama dengan jumlah baris pada collection `movies`. Untuk mewujudkan hal tersebut, maka dapat dengan melakukan komperasi imdb list antar collection dengan menggunakan fungsi `$match`. Sebelumnya, perlu memastikan imdb_liss menjadi bentuk array.

In [13]:
# menjadikan imdb keys berbentuk array
for i in imdb:
  imdb_list.append (list(i.values())[0])
  
imdb_list[:5]

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

### 1.3 Make the Pipeline
> Pipeline dibentuk sesuai dengan karakteristik dua collection (collection `movies` dan `movies_initial`)yang telah diketahui pada tahap sebelumnya. 


In [14]:
pipeline = [
    {
        '$project' : {
            # to list
            'awards': {
                'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                'text': "$awards",
                'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            # cast = cast (same)
            'cast' : {"$cond": [{"$eq": ["$cast", ""]}, "$nonExistinField", {'$split' : ['$cast', ', ']}]},
            # countries = country
            'countries' : {"$cond": [{"$eq": ["$country", ""]}, "$nonExistinField", {'$split' : ['$country', ', ']}]},   
            # directors = director
            'directors' : {"$cond": [{"$eq": ["$director", ""]}, "$nonExistinField", {'$split' : ['$director', ', ']}]},    
            # fullplot = fullplot (same)
            'fullplot': {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", "$fullplot"]},
            # genres = genre
            'genres': {"$cond": [{"$eq": ["$genre", ""]}, "$nonExistinField", {'$split' : ['$genre', ', ']}]},
            # many column to one column (list)
            'imdb' : {
                'id' : '$imdbID',
                'rating' : '$imdbRating',
                'votes' : '$imdbVotes',
                },
            # languages = language
            'languages': {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", "$language"]},
            # lastupdated = lastupdated (same)
            'lastupdated': {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", "$lastupdated"]},    
            # metacritic =  metacritic (same)            
            'metacritic': {"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},   
            # tidak terdapat num_mflix_comments pada movies_initial
            'num_mflix_comments' : " ",
            # plot = plot (same)
            'plot': {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},   
            # poster = poster (same)
            'poster': {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]}, 
            # rated = rating
            'rated': {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", "$rating"]},
            # string to datetime format
            'released': {"$cond": [{"$eq": ["$released", ""]}, "$nonExistinField", 
                                   {'$cond': {'if': {'$ne': ["$released", ""]},
                                   'then': {
                                       '$dateFromString': {
                                           'dateString': "$released"
                                       }
                                   },
                                   'else': ""}
                        }]} ,
            # hanya mengambil angka saja (sebagian value)
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, "$nonExistinField" ,                               
                                  {'$arrayElemAt': [{'$split': [ "$runtime", " min" ]}, 0]}]},
            # title = title (same)
            'title': {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},      
            # tidak terdapat tomatoes pada movies_initial
            'tomatoes' : {
                'lastUpdated' : '',
                'viewer' : {'numReviews' : '', 'rating' : ''},
                },          
            # type = type (same)
            'type': {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]}, 
            # writers = writer
            'writers': {"$cond": [{"$eq": ["$writer", ""]}, "$nonExistinField", {'$split' : ['$writer', ', ']}]},  
            # year = year (same)
            'year': {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]}                      
        }
    },
    {
        # menjadikan jumlah baris sama dengan value untuk setiap document sesuai
        '$match':{'imdb':{'$in':
                          [dict(sorted(imdb_list[x].items(), key=lambda x: x[0]))
                           for x in range(len(imdb_list))]}}  
    },
    {
        # membuat collecton baru sebagai output
        '$out' : 'clean_movies_afifah' 
    }
]

clear_output()
pprint.pprint(list(db0.movies_initial.aggregate(pipeline)))

[]


### 1.3 Check the New Collection 
Untuk memastikan apakah collection baru (`clean_movies_afifah`) yang telah dibentuk berdasarkan pipeline diatas telah terbentuk, maka dapat dilihat pada output berikut.


In [17]:
pprint.pprint(db0.clean_movies_afifah.find_one())

{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': ' ',
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': datetime.d

> Dengan melakukan running pada query di atas, maka dapat diketahui bahwa collection baru telah terbentuk. Untuk mengetahui apakah collection yang dibentuk telah sesuai, maka dapat melanjukannya pada tahab berikutnya untuk melakukan validasi.

## LANGKAH 2 :Validasi collection 
<br>
Berdasarkan penjelasan di awal, maka dapat diketahui terdapat empat parameter untuk melakukan validasi collection yang telah dibuat. Maka dari tu akan dilakukan penjelasan untuk setiap parameter dengan menjadikannya tiga poin. Parameter yang pertama dan kedua akan dibahas dalam poin satu, parameter ketiga akan dibahas dalam poin 2 dan selanjutnya parameter keempat akan dibahas dalam poin 3.

### 2.1 Semua Document dan Buanyak Document pada `clean_movies` dan `movie` sama

> Jika ingin mengtahui apakah semua document sama, maka dapat dipastikan jumlah document adalah sama. Maka dari itu parameter satu dan dua akan dibahas dalam satu poin. 
>
> Jika dilihat pada bagian 1.2 maka dapat diketahui bahwa jumlah baris pada `clean_movies_afifah` (collecton baru) seharusnya sebanyak 23541 baris. Hal tersebut dapan kembali dimunculkan sebagai berikut. Dan dapat dibandingkan dengan jumlah baris pada collection `movies`.



In [140]:
print ('Jumlah baris movies :', len(imdb))

clean_movies = list(db0.clean_movies_afifah.find({}))
print ('Jumlah baris clean_movies_afifah :', len(clean_movies))

Jumlah baris movies : 23541
Jumlah baris clean_movies_afifah : 23539


> Jika membandingkan kedua query di atas, maka dapat diketahui terdapat perbedaan jumlah baris antara collection `clean_movies_afifah` dan collection `movies`. Jumlah baris collection `clean_movies_afifah` lebih sedikit (23.539) jika dibandingkan dengan jumlah baris pada collection `movie` (23.541). Terdapat dua baris lebih banyak pada collection `movie`. Seharusnya hal tersebut tidak terjadi, dikarenakan pada pipeline telah menggunakan fungsi `'$match'` untuk melakukan agregasi antara collection `movie_initial` dan `movie`. 
<br>
<br>
Untuk mengetahui penyebab jumlah baris yang dihasilkan tidak sama, dapat melakukan running query sebagai berikut.

In [84]:
grup_clean = list(db0.clean_movies_afifah.aggregate([{"$sortByCount":"$imdb"}]))
grup_movies = list(client0.sample_mflix.movies.aggregate([{"$sortByCount":"$imdb"}]))

for i in grup_movies:
  if i not in grup_clean:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


> Berdasarkan query di atas, maka dapat mendetksi id IMDB pada collection `movie` yang tidak terdapat pada collection `clean_movies_afifah`. Dapat diketahui IMDB id 5 dengan rating 6.2 dan votes 1189 tidak muncul tiga kali pada collection `clean_movies_afifah`. Untuk mengetahui document dengan IMDB id = 5 pada `movie` dan `clean_movies_afifah` dapat dilihat pada data frame berikut   

In [134]:
# IMDB id = 5 pada movie
pd.DataFrame(list(col01.find({'imdb.id':{'$in':[5]}})))

_id  ...  year
0  573a1390f29313caabcd4135  ...  1893
1  5dafd0d3d75e5fa8b23dae21  ...  1893
2  5dafd159d75e5fa8b23dae22  ...  1893

[3 rows x 18 columns]

In [135]:
# IMDB id = 5 pada clean_movies_afifah
pd.DataFrame(list(db0.clean_movies_afifah.find({'imdb.id':{'$in':[5]}})))

_id  ...  year
0  5daf3e85ab0f2781ea55d596  ...  1893

[1 rows x 18 columns]

> Output di atas menunjukkan bahwa judul movie yang memikiliki IMDB id sama dengan 5 pada collection `movie` adalah *Blacksmith Scene* dan *Blacksmith Scene II*. Jika ditinjau lebih lanjut, ketiga document memiliki value yang sama pada setiap keys kecuali jika dilihat pada keys title-nya. Hal tersebut dapat diduga, adanya salah imputasi data pada collection `movie`. Hal tersebut diduga karena ketiga title tidak jauh berbeda dan dapat dikatakan adanya informasi yang berulang pada collection `movie`. Maka dapat diketahui bahwa dengan berbedanya jumlah baris pada kedua collection tidak menjadi masalah. Hal tersebut dapat dipastikan bahwa keseluruhan document pada collection `clean_movies_afifah` tidak mengandung value yang sama.
<br>

**Catatan :** <br>
Semua Document dan Buanyak Document pada clean_movies dan movie tidak sama dikarenakan terdapat `imdb_list` yang berulang pada collection `movies`. Sehingga collection `clean_movies_afifah` terdeteksi lebih sedikit jika dibandingkan dengan collection `movies`. Walaupu document dan banyak document antar collection tidak sama, namun dapat dipastikan tidak terdapat value dengan informasi yang sama pada collection `clean_movies_afifah`.


### 2.2 Semua fields pada `clean_movies` ada pada `movie`
Untuk mengetahui apakah fields pada collection `clean_movies_afifah` ada pada collection `movie` dapat 

In [153]:
key_movies

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [155]:
# how to show all keys MOVIES in one vertical list
key_clean = [list(doc.keys()) 
              for doc in db0.clean_movies_afifah.find({})]
key_clean = set([j 
                  for i in key_clean
                  for j in i])
key_clean

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [156]:
print ('Jumlah keys/fields pada  movies :', len(key_movies))
print ('Jumlah keys/fields pada clean_movies_afifah :', len(key_clean))

Jumlah keys/fields pada  movies : 22
Jumlah keys/fields pada clean_movies_afifah : 22


> Dengan memunculkan keseluruhan keys pada collection `movies` dan `clean_movies_afifah`, dan menghitung jumlah keys pada dua collection tersebut maka dapat dipastikan semua fields atau keys pada `clean_movies` atau `clean_movies_afifah` ada pada collection `movies`.

### 2.3 Semua value pada `clean_movies` sama dengan semua value pada `movies` dengan urutan yang sama

Untuk mengetahui apakah semua document antara kedua collection sama, dapat melihat pada salah satu Title nya dan dibandingkan documentnya. 


In [98]:
# movies
pprint.pprint(list(db0.movies.find({'title':'Blacksmith Scene'})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

In [127]:
# clean_movies_afifah
pprint.pprint(list(client0.sample_mflix.clean_movies_afifah.find({'title':'Blacksmith Scene'})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': ' ',
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  're

In [123]:
# movies
pd.DataFrame(list(db0.movies.find({'title':'Blacksmith Scene'}))) 

_id  ...  year
0  573a1390f29313caabcd4135  ...  1893

[1 rows x 18 columns]

In [128]:
# clean movies
pd.DataFrame(list(client0.sample_mflix.clean_movies_afifah.find({'title':'Blacksmith Scene'}))) 

_id  ...  year
0  5daf3e85ab0f2781ea55d596  ...  1893

[1 rows x 18 columns]

> Jika dilihat berdasarkan keseluruhan output find dengan melihat document bersayarat `'title':'Blacksmith Scene'`, maka dapat diketahui bahwa keys dan value document tersebut sama jika membandingkan document pada kedua collection tersebut (`clean_movies_afifah` vs `movies`).

> Untuk memastikannya lebih lanjut, maka dapat dilakukan pengecekan pada title yang lain, sebagai berikut.

In [19]:
# movies
pprint.pprint(list(db0.movies.find({'title':'Titanic'})))

[{'_id': ObjectId('573a1394f29313caabcdf639'),
  'awards': {'nominations': 3,
             'text': 'Won 1 Oscar. Another 2 nominations.',
             'wins': 0},
  'cast': ['Clifton Webb',
           'Barbara Stanwyck',
           'Robert Wagner',
           'Audrey Dalton'],
  'countries': ['USA'],
  'directors': ['Jean Negulesco'],
  'fullplot': 'Unhappily married and uncomfortable with life among the British '
              'upper crust, Julia Sturges takes her two children and boards '
              'the Titanic for America. Her husband Richard also arranges '
              'passage on the doomed luxury liner in order to let him have '
              'custody of their two children. Their problems soon seem minor '
              'when the ship hits an iceberg.',
  'genres': ['Drama', 'History', 'Romance'],
  'imdb': {'id': 46435, 'rating': 7.3, 'votes': 4677},
  'languages': ['English', 'Basque', 'French', 'Spanish'],
  'lastupdated': '2015-09-16 00:00:16.593000000',
  'num_mflix_co

In [20]:
# clean_movies_afifah
pprint.pprint(list(client0.sample_mflix.clean_movies_afifah.find({'title':'Titanic'})))

[{'_id': ObjectId('5daf3e89ab0f2781ea55e5d2'),
  'awards': {'nominations': '2',
             'text': 'Won 1 Oscar. Another 2 nominations.',
             'wins': None},
  'cast': ['Clifton Webb',
           'Barbara Stanwyck',
           'Robert Wagner',
           'Audrey Dalton'],
  'countries': ['USA'],
  'directors': ['Jean Negulesco'],
  'fullplot': 'Unhappily married and uncomfortable with life among the British '
              'upper crust, Julia Sturges takes her two children and boards '
              'the Titanic for America. Her husband Richard also arranges '
              'passage on the doomed luxury liner in order to let him have '
              'custody of their two children. Their problems soon seem minor '
              'when the ship hits an iceberg.',
  'genres': ['Drama', 'History', 'Romance'],
  'imdb': {'id': 46435, 'rating': 7.3, 'votes': 4677},
  'languages': 'English, Basque, French, Spanish',
  'lastupdated': '2015-09-16 00:00:16.593000000',
  'num_mflix_comme

> Melakukan pengecekan ulang terhadap title Titanic, maka dapat diketahui pada saat keys yang tidak ada pada document sebelumnya akan dapat dimuculkan pada documen yang memiliki values pada keys tersebut. 

### KESIMPULAN 
Practice 1 mengenai NoSQL pada PyMongo dalam membuat collection baru ini menunjukkan bahwa collection baru yang terbentuk (`clean_movies_afifah`) memiliki jumlah yang berbeda jika dibandingkan dengan collection target (`movies`). Namun dapat dipastikan tidak terdapat values yang berulang. Dan dikarenakan tidak terdapat `num_mflix_comments` dan `tomatos` ada collection `movies_initial`, sedangkan kedua keys tersebut ada pada `movies`, maka pada `clean_movies_afifah` kedua keys tersebut tidak memiliki values. Namun unjuk jumlah keys pada `clean_movies_afifah` sudah sama persis dengan `movies`.